src/tools/w3dviewer/mainfrm.cpp

```diff
@@ -0,0 +1,1616 @@
+    ShowControlBar(&m_animationToolbar, FALSE, FALSE);
+
+    m_statusBar.SetPaneText(1, "");
+    m_statusBar.SetPaneText(2, "");
+    m_statusBar.SetPaneText(3, "");
+    m_statusBar.SetPaneText(4, "");
+    m_statusBar.SetPaneText(5, "");
+    m_statusBar.SetPaneText(6, "");
+
+    W3DAssetManager::Get_Instance()->Set_W3D_Load_On_Demand(true);
+    W3DAssetManager::Get_Instance()->Set_Activate_Fog_On_Load(true);
+    GetWindowRect(&m_windowRect);
+    HMENU menu = ::GetMenu(m_hWnd);
+    HMENU SubMenu = GetSubMenu(menu, 3);
+    m_subMenu = SubMenu;
+    MENUITEMINFO info;
+    memset(&info, 0, sizeof(info));
+    info.cbSize = sizeof(info);
+    info.fMask = MIIM_STRING;
+    char buffer[200];
+    info.dwTypeData = buffer;
+    info.cch = 100;
+    GetMenuItemInfo(SubMenu, 4, TRUE, &info);
+    m_subMenu = GetSubMenu(SubMenu, 4);
+    RestoreWindowPos();
+    m_initialized = TRUE;
+    return (Device != -1) - 1;
+}
+
+void CMainFrame::OnActivateApp(BOOL bActive, DWORD dwThreadID)
+{
+    CGraphicView *view = reinterpret_cast<CGraphicView *>(m_splitter.GetPane(0, 1));
+
+    if (view != nullptr) {
+        view->m_isActive = bActive;
+
+        if (!bActive) {
+            SetProp(m_hWnd, "Inactive", (HANDLE)1);
+            CWnd::Default();
+            return;
+        }
+
+        RemoveProp(m_hWnd, "Inactive");
+        view->m_time = GetTickCount();
+    }
+
+    CWnd::Default();
+    CWnd::Default();
+}
+
+void CMainFrame::OnWindowPosChanging(LPWINDOWPOS lpWndPos)
+{
+    CWnd::Default();
+}
+
+void CMainFrame::OnWindowPosChanged(WINDOWPOS FAR *lpwndpos)
+{
+    CWnd::Default();
+}
+
+void CMainFrame::OnDestroy()
+{
+    WINDOWPLACEMENT placement;
+    memset(&placement, 0, sizeof(placement));
+    placement.length = sizeof(WINDOWPLACEMENT);
+    GetWindowPlacement(&placement);
+    RECT r;
+    GetWindowRect(&r);
+    theApp.WriteProfileInt("Window", "Left", r.left);
+    theApp.WriteProfileInt("Window", "Right", r.right);
+    theApp.WriteProfileInt("Window", "Top", r.top);
+    theApp.WriteProfileInt("Window", "Bottom", r.bottom);
+    theApp.WriteProfileInt("Window", "Maximized", placement.showCmd == SW_SHOWMAXIMIZED);
+    theApp.WriteProfileInt("Config", "AnimateCamera", reinterpret_cast<CW3DViewDoc *>(GetActiveDocument())->m_animateCamera);
+    theApp.WriteProfileInt("Config", "ResetCamera", reinterpret_cast<CW3DViewDoc *>(GetActiveDocument())->m_resetCamera);
+    CFrameWnd::OnDestroy();
+}
+
+void CMainFrame::OnProperties()
+{
+    // do later
+}
+
+void CMainFrame::OnGenerateLOD()
+{
+    // do later
+}
+
+void CMainFrame::OnOpen()
+{
+    CW3DViewDoc *doc = reinterpret_cast<CW3DViewDoc *>(GetActiveDocument());
+
+    if (doc != nullptr) {
+        CFileDialog dlg(TRUE,
+            ".w3d",
+            nullptr,
+            OFN_EXPLORER | OFN_ALLOWMULTISELECT | OFN_HIDEREADONLY | OFN_OVERWRITEPROMPT,
+            "Westwood 3D Files (*.w3d)|*.w3d||",
+            this);
+        char file[2600] = {};
```

Comment:
This buffer is 2600 ? I assume because user can write multiple chained file paths maybe?

Reply:
Yes.